File: problem-solving/lc-746-easy-min-cost-climbing-stairs/2-bottom-up.py
"""
https://leetcode.com/problems/min-cost-climbing-stairs/

Complexity
----------
Time: O(N)
Space: O(N)
"""

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n = len(cost)
        dp = [0 for _ in range(n + 1)]
        for idx in range(2, n + 1):
            dp[idx] = min(dp[idx - 1] + cost[idx - 1],
                          dp[idx - 2] + cost[idx - 2])
        return dp[n]

File: problem-solving/_lc-0115-hard-distinct-subsequences/2-bottom-up.go
/*
https://leetcode.com/problems/distinct-subsequences/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
*/

func numDistinct(s string, t string) int {
    m, n := len(s), len(t)
    var dp [][]int

    for i := 0; i < m + 1; i++ {
        dp = append(dp, make([]int, n + 1))
    }

    for i := 0; i < m + 1; i++ {
        dp[i][0] = 1
    }

    for i := 1; i < m + 1; i++ {
        for j := 1; j < n + 1; j++ {
            if s[i - 1] == t[j - 1] {
                dp[i][j] = dp[i - 1][j] + dp[i - 1][j - 1]
            } else {
                dp[i][j] = dp[i - 1][j]
            }
        }
    }
    return dp[m][n]
}
